File: app/src/main/java/me/magnum/melonds/ui/romlist/RomConfigDialog.java
package me.magnum.melonds.ui.romlist;

import android.content.Context;
import android.os.Bundle;
import android.os.Environment;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.DialogTitle;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.Switch;
import android.widget.TextView;
import com.github.angads25.filepicker.controller.DialogSelectionListener;
import com.github.angads25.filepicker.model.DialogConfigs;
import com.github.angads25.filepicker.model.DialogProperties;
import com.github.angads25.filepicker.view.FilePickerDialog;
import me.magnum.melonds.R;
import me.magnum.melonds.model.RomConfig;
import me.magnum.melonds.utils.UIUtils;

import java.io.File;

public class RomConfigDialog extends AlertDialog {
    public interface OnRomConfigSavedListener {
        void onRomConfigSaved(RomConfig romConfig);
    }

    private String title;
    private RomConfig romConfig;
    private OnRomConfigSavedListener saveListener;

    private View prefLoadGbaRomView;
    private View prefGbaRomPathView;
    private View prefGbaSavePathView;
    private Switch loadGbaRomSwitch;
    private TextView gbaRomPathTextView;
    private TextView gbaSavePathTextView;

    public RomConfigDialog(@NonNull Context context, String title, RomConfig romConfig) {
        super(context);
        this.title = title;
        this.romConfig = romConfig;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.dialog_rom_config);
        setCancelable(true);

        this.prefLoadGbaRomView = this.findViewById(R.id.layout_rom_config_load_gba_rom);
        this.prefGbaRomPathView = this.findViewById(R.id.layout_rom_config_gba_rom_path);
        this.prefGbaSavePathView = this.findViewById(R.id.layout_rom_config_gba_save_path);
        this.loadGbaRomSwitch = this.findViewById(R.id.switch_load_gba_rom);
        this.gbaRomPathTextView = this.findViewById(R.id.label_preference_gba_cart_path_value);
        this.gbaSavePathTextView = this.findViewById(R.id.label_preference_gba_save_path_value);

        this.prefLoadGbaRomView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                loadGbaRomSwitch.setChecked(!loadGbaRomSwitch.isChecked());
            }
        });
        this.prefGbaRomPathView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DialogProperties properties = new DialogProperties();
                properties.selection_mode = DialogConfigs.SINGLE_MODE;
                properties.selection_type = DialogConfigs.FILE_SELECT;
                properties.root = Environment.getExternalStorageDirectory();
                FilePickerDialog pickerDialog = new FilePickerDialog(getContext(), properties);
                pickerDialog.setDialogSelectionListener(new DialogSelectionListener() {
                    @Override
                    public void onSelectedFilePaths(String[] files) {
                        if (files.length > 0) {
                            if (new File(files[0]).isFile())
                                onGbaRomPathSelected(files[0]);
                        }
                    }
                });
                pickerDialog.show();
            }
        });
        this.prefGbaSavePathView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DialogProperties properties = new DialogProperties();
                properties.selection_mode = DialogConfigs.SINGLE_MODE;
                properties.selection_type = DialogConfigs.FILE_SELECT;
                properties.root = Environment.getExternalStorageDirectory();
                FilePickerDialog pickerDialog = new FilePickerDialog(getContext(), properties);
                pickerDialog.setDialogSelectionListener(new DialogSelectionListener() {
                    @Override
                    public void onSelectedFilePaths(String[] files) {
                        if (files.length > 0) {
                            if (new File(files[0]).isFile())
                                onGbaSavePathSelected(files[0]);
                        }
                    }
                });
                pickerDialog.show();
            }
        });
        this.loadGbaRomSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                setLoadGbaRom(isChecked);
            }
        });

        ((DialogTitle) this.findViewById(R.id.text_rom_config_title)).setText(title);
        this.loadGbaRomSwitch.setChecked(romConfig.loadGbaCart());
        this.gbaRomPathTextView.setText(getPathOrDefault(romConfig.getGbaCartPath()));
        this.gbaSavePathTextView.setText(getPathOrDefault(romConfig.getGbaSavePath()));
        UIUtils.setViewEnabled(this.prefGbaRomPathView, romConfig.loadGbaCart());
        UIUtils.setViewEnabled(this.prefGbaSavePathView, romConfig.loadGbaCart());

        this.findViewById(R.id.button_rom_config_ok).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (saveListener != null)
                    saveListener.onRomConfigSaved(romConfig);
                dismiss();
            }
        });
        this.findViewById(R.id.button_rom_config_cancel).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dismiss();
            }
        });
    }

    public RomConfigDialog setOnRomConfigSaveListener(OnRomConfigSavedListener listener) {
        this.saveListener = listener;
        return this;
    }

    private void setLoadGbaRom(boolean loadGbaRom) {
        this.romConfig.setLoadGbaCart(loadGbaRom);
        UIUtils.setViewEnabled(this.prefGbaRomPathView, loadGbaRom);
        UIUtils.setViewEnabled(this.prefGbaSavePathView, loadGbaRom);
    }

    private void onGbaRomPathSelected(String romPath) {
        this.romConfig.setGbaCartPath(romPath);
        this.gbaRomPathTextView.setText(getPathOrDefault(romPath));
    }

    private void onGbaSavePathSelected(String savePath) {
        this.romConfig.setGbaSavePath(savePath);
        this.gbaSavePathTextView.setText(getPathOrDefault(savePath));
    }

    private String getPathOrDefault(String path) {
        if (path != null)
            return path;

        return getContext().getString(R.string.not_set);
    }
}

File: app/src/main/java/me/magnum/melonds/ui/romlist/RomConfigDialog.kt
package me.magnum.melonds.ui.romlist

import android.Manifest
import android.app.Activity
import android.app.Dialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.core.os.bundleOf
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.activityViewModels
import io.reactivex.disposables.Disposable
import me.magnum.melonds.R
import me.magnum.melonds.common.Permission
import me.magnum.melonds.databinding.DialogRomConfigBinding
import me.magnum.melonds.domain.model.Rom
import me.magnum.melonds.domain.model.RomConfig
import me.magnum.melonds.domain.model.RuntimeConsoleType
import me.magnum.melonds.domain.model.RuntimeMicSource
import me.magnum.melonds.extensions.setViewEnabledRecursive
import me.magnum.melonds.parcelables.RomConfigParcelable
import me.magnum.melonds.parcelables.RomParcelable
import me.magnum.melonds.ui.layouts.LayoutSelectorActivity
import me.magnum.melonds.common.contracts.FilePickerContract
import me.magnum.melonds.extensions.isMicrophonePermissionGranted
import me.magnum.melonds.utils.FileUtils
import java.util.*

class RomConfigDialog : DialogFragment() {
    companion object {
        private const val KEY_TITLE = "title"
        private const val KEY_ROM = "rom"
        private const val KEY_ROM_CONFIG = "rom_config"

        fun newInstance(title: String, rom: Rom): RomConfigDialog {
            return RomConfigDialog().apply {
                arguments = bundleOf(
                    KEY_TITLE to title,
                    KEY_ROM to RomParcelable(rom)
                )
            }
        }
    }

    private val romListViewModel: RomListViewModel by activityViewModels()
    private lateinit var binding: DialogRomConfigBinding
    private var layoutNameDisposable: Disposable? = null

    private lateinit var rom: Rom
    private lateinit var romConfig: RomConfig

    private val layoutPickerLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            val layoutId = result.data?.getStringExtra(LayoutSelectorActivity.KEY_SELECTED_LAYOUT_ID)?.let { UUID.fromString(it) }
            onLayoutIdSelected(layoutId)
        }
    }
    private val gbaRomFilePicker = registerForActivityResult(FilePickerContract(Permission.READ)) {
        if (it != null) {
            onGbaRomPathSelected(it)
        }
    }
    private val gbaSramFilePicker = registerForActivityResult(FilePickerContract(Permission.READ_WRITE)) {
        if (it != null) {
            onGbaSavePathSelected(it)
        }
    }
    private val microphonePermissionLauncher =
    registerForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
        if (granted) {
            onRuntimeMicSourceSelected(RuntimeMicSource.DEVICE)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // ROM is immutable so we can use the arguments' one. Only the ROM config needs to be saved if the fragment is rebuilt
        rom = arguments?.getParcelable<RomParcelable>(KEY_ROM)?.rom ?: return
        romConfig = if (savedInstanceState != null) {
            savedInstanceState.getParcelable<RomConfigParcelable>(KEY_ROM_CONFIG)?.romConfig ?: return
        } else {
            rom.config.copy()
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        binding = DialogRomConfigBinding.inflate(layoutInflater)

        return AlertDialog.Builder(requireContext())
                .setView(binding.root)
                .setCancelable(true)
                .create()
    }

    override fun onStart() {
        super.onStart()
        isCancelable = true

        val title = arguments?.getString(KEY_TITLE)

        binding.layoutPrefSystem.setOnClickListener {
            AlertDialog.Builder(requireContext())
                    .setTitle(R.string.label_rom_config_console)
                    .setSingleChoiceItems(R.array.game_runtime_console_type_options, romConfig.runtimeConsoleType.ordinal) { dialog, which ->
                        val newConsoleType = RuntimeConsoleType.values()[which]
                        onRuntimeConsoleTypeSelected(newConsoleType)
                        dialog.dismiss()
                    }
                    .setNegativeButton(R.string.cancel) { dialog, _ ->
                        dialog.cancel()
                    }
                    .show()
        }
        binding.layoutPrefRuntimeMicSource.setOnClickListener {
            AlertDialog.Builder(requireContext())
                    .setTitle(R.string.microphone_source)
                    .setSingleChoiceItems(R.array.game_runtime_mic_source_options, romConfig.runtimeMicSource.ordinal) { dialog, which ->
                        val newMicSource = RuntimeMicSource.values()[which]
                        // Request mic permission if required
                        if (newMicSource == RuntimeMicSource.DEVICE && !requireContext().isMicrophonePermissionGranted()) {
                            microphonePermissionLauncher.launch(Manifest.permission.RECORD_AUDIO)
                        } else {
                            onRuntimeMicSourceSelected(newMicSource)
                        }

                        dialog.dismiss()
                    }
                    .setNegativeButton(R.string.cancel) { dialog, _ ->
                        dialog.cancel()
                    }
                    .show()
        }
        binding.layoutPrefLayout.setOnClickListener {
            val intent = Intent(requireContext(), LayoutSelectorActivity::class.java).apply {
                putExtra(LayoutSelectorActivity.KEY_SELECTED_LAYOUT_ID, romConfig.layoutId?.toString())
            }
            layoutPickerLauncher.launch(intent)
        }
        binding.layoutPrefLoadGbaRom.setOnClickListener { binding.switchLoadGbaRom.toggle() }
        binding.layoutPrefGbaRomPath.setOnClickListener {
            gbaRomFilePicker.launch(Pair(romConfig.gbaCartPath, null))
        }
        layoutNameDisposable = romListViewModel.getLayout(romConfig.layoutId).subscribe { layout ->
            binding.textPrefLayout.text = layout.name
        }
        binding.layoutPrefGbaSavePath.setOnClickListener {
            gbaSramFilePicker.launch(Pair(romConfig.gbaSavePath, null))
        }
        binding.switchLoadGbaRom.setOnCheckedChangeListener { _, isChecked -> setLoadGbaRom(isChecked) }
        binding.textRomConfigTitle.text = title

        onRuntimeConsoleTypeSelected(romConfig.runtimeConsoleType)
        onRuntimeMicSourceSelected(romConfig.runtimeMicSource)

        binding.switchLoadGbaRom.isChecked = romConfig.loadGbaCart
        binding.textPrefGbaRomPath.text = getUriPathOrDefault(romConfig.gbaCartPath)
        binding.textPrefGbaSavePath.text = getUriPathOrDefault(romConfig.gbaSavePath)

        binding.layoutPrefGbaRomPath.setViewEnabledRecursive(romConfig.loadGbaCart)
        binding.layoutPrefGbaSavePath.setViewEnabledRecursive(romConfig.loadGbaCart)

        binding.buttonRomConfigOk.setOnClickListener {
            romListViewModel.updateRomConfig(rom, romConfig)
            dismiss()
        }
        binding.buttonRomConfigCancel.setOnClickListener { dismiss() }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(KEY_ROM_CONFIG, RomConfigParcelable(romConfig))
    }

    private fun setLoadGbaRom(loadGbaRom: Boolean) {
        romConfig.loadGbaCart = loadGbaRom
        binding.layoutPrefGbaRomPath.setViewEnabledRecursive(loadGbaRom)
        binding.layoutPrefGbaSavePath.setViewEnabledRecursive(loadGbaRom)
    }

    private fun onRuntimeConsoleTypeSelected(consoleType: RuntimeConsoleType) {
        val options = resources.getStringArray(R.array.game_runtime_console_type_options)
        romConfig.runtimeConsoleType = consoleType
        binding.textPrefRuntimeConsoleType.text = options[consoleType.ordinal]
    }

    private fun onRuntimeMicSourceSelected(micSource: RuntimeMicSource) {
        val options = resources.getStringArray(R.array.game_runtime_mic_source_options)
        romConfig.runtimeMicSource = micSource
        binding.textPrefRuntimeMicSource.text = options[micSource.ordinal]
    }

    private fun onLayoutIdSelected(layoutId: UUID?) {
        romConfig.layoutId = layoutId
        layoutNameDisposable?.dispose()
        layoutNameDisposable = romListViewModel.getLayout(layoutId).subscribe { layout ->
            binding.textPrefLayout.text = layout.name
        }
    }

    private fun onGbaRomPathSelected(romFileUri: Uri) {
        romConfig.gbaCartPath = romFileUri
        binding.textPrefGbaRomPath.text = getUriPathOrDefault(romFileUri)
    }

    private fun onGbaSavePathSelected(saveFileUri: Uri) {
        romConfig.gbaSavePath = saveFileUri
        binding.textPrefGbaSavePath.text = getUriPathOrDefault(saveFileUri)
    }

    private fun getUriPathOrDefault(uri: Uri?): String {
        return FileUtils.getAbsolutePathFromSAFUri(requireContext(), uri) ?: getString(R.string.not_set)
    }

    override fun onStop() {
        super.onStop()
        layoutNameDisposable?.dispose()
    }
}